### DISCUSSION TITLE
Best way to create function that depends on additional variable

### DISCUSSION THREAD

**Discussion by gagamail** [imk-toolkit/imk-toolkit#44] [Q&A] [answered]
What is the best way to create a function that depends on another additional variable?

As example I use a modification of anomalies.py:
Just imagine instead of calculating the anomaly to the climatology of the dataarray I want to calculate the anomaly to an external choosable climatology (of course already in the same format as dataarray). How would I do this with the given restrictions of the toolkit?

**Comment by ucyo** [comment, marked as answer]
Hi @gagamail,

what do you mean with an externally choosable climatology? If there are any dependencies, you could give it as a parameter to method.
Lets say we want to change the `anomalies` function in such a way, that we want to add an offset to the results anomalies with an offset.
Then the `main` implementation could look like the following:

```python
def main(dataarray, offset):  # changed number of arguments
    """Return monthly anomalies for dataarray."""
    assert isinstance(dataarray, xr.DataArray), "Input is not an xr.DataArray"
    climatology = dataarray.groupby("time.month").mean("time")
    anomalies = dataarray.groupby("time.month") - climatology
    
    # Added changed
    print(f"Adding offset {offset} to anomalies") 
    anomalies -= offset               

    return anomalies
```
Then, the example given in the [README](https://github.com/imk-toolkit/imk-toolkit#usage) would look like:
```python
import imktk
import xarray as xr

offset = 4
t = xr.tutorial.open_dataset("rasm").load().Tair
anomaly_t_with_offset = t.imktk.anomalies(offset)
```
The actual DataArray `t` is given as the first argument to the new `anomalies.main` function and all the remaining parameter are given as described in the `t.imktk.anomalies(offset)` call. Another way to put it, is that when you execute `<var>.imktk.<scriptname>(*args, *kwargs)`, then - what is actually being executed is -  `main(var, *args, **kwargs)` in `<scriptname>.py`.
